### DISCUSSION TITLE
Multi-tenancy using the same database

### DISCUSSION THREAD

**Discussion by akiarostami** [directus/directus#4832] [Feature Requests]
I've been chatting with you guys in [Discord](https://discord.com/channels/725371605378924594/761221022216093737/827591482226769961) about this (and you are amazing!) Per your request, here's the summary of our discussion:

Right now there are two solutions for multi-tenancy:

- using separate databases: this is a clean solution and I like that it keeps every client's data completely separate from the others. However, if we want to have the same collection for all tenants (which is the case for me), it would be difficult to keep all these databases in synced.
- the other option is achieving multi-tenancy through roles and filtering data based on $CURRENT_USER. The problem with that is I lose the ability to have multiple users for each tenant, as only the owner of the records would be able to access the data, and not their teammates.

I believe there is a simple solution for this: using Project ID. For this, we would have to do two things:

- In collections, the same way we have "created by" and "created on", we should be able to add "project id" to be automatically added to every record created in collections. As Rijk suggested, this can part also be achieved by an extension:

`module.exports = function() {
  return {
    'items.create.before': async (payload) => ({
      ...payload,
      project: process.env.UNIQUE_PROJECT_ID
    })
  }
}`

- We would also need to be able to use this in the filters. The easiest way is to be able to access the UNIQUE_PROJECT_ID as the environment variable such as $CURRENT_PROJECT. Ideally, one should be able to access other variables in the filters as well, but that's a different discussion. :)

I believe this would address the first level multi-tenancy using the same database. There will be a problem though: all tenants would have the same logo / colors / settings. But for users who are OK with having the same logo  and project settings (like me!), this would address all their needs.

The reason for wanting to have the same database in my case is this: 
I am working on an idea for hosing a publishing platform with specific set of collections. All my (potential) users would use those collections. Imagine we have 1K users, and we want to make a change in one collection. Or add a new collection to the platform. We would have to update 1000 databases. 
Also, Most users won't have tens of thousands of records. It would be much better and cheaper to have one big powerful database than 1000 smaller databases.
Lastly, this way it would be easier to get system-wide analytics.

**Comment by rijkvanzanten** [comment]
Note(s) to self:

For this to work as expected:

* We'd have to expose the project's unique key based on the environment variable, so the app is aware and can pre-fill fields based on these presets
* We should add a dynamic var for this projects unique key for filtering / permissions purposes
* We should add "set to current project id" as a on-create/on-update flag*
* We should allow permissions to control what singular item is used for singleton collections. This would allow things like directus_settings to be "scoped" based on the project ID

**Reply by Ataraxy** [reply]
Similarly, but different, I share this same issue of desiring a single database / multi tenant / multi user project.

Example use case for this would be a multi tenant customer service platform where third party agents are answering messages on behalf of all tenants but tenants should still only have access to their own data. Tenants also need to be able to have sub users which also only have access to the parent tenants data.

I think one straight forward way to solve this is by adding a tenant_id column to a user record and then creating a role filter rule that would check for `$CURRENT_USER` or a potential `$CURRENT_USER_RECORD.tenant_id` but would need the current user record exposed to filter rules to be able to do this.

**Comment by rijkvanzanten** [comment]
Heya! Thanks for opening this feature request!

This feature request has received over **15 votes** from the community. This means we'll move this feature request to the Under Review state! 

The Core team will schedule a meeting to review this request as soon as possible. The discussion will then be approved or denied. You may or may not be invited to join this meeting with the core team.

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._

**Comment by Zyles** [comment]
I am looking for multi tenancy using one nodejs and one postgresql installation with one tenant database per project.

Now I would need to deploy one installation of Directus per project instead of horizontally scaling one installation.

**Comment by DanielBiegler** [comment]
Just dropping a resource that might help somebody looking into this: In our recent Meetup there was a talk dedicated to multitenancy! See here https://www.youtube.com/watch?v=RU-kAVp9aH4

**Reply by bbbjames** [reply]
thank you Daniel 🙏

also the article https://hybridheroes.de/blog/2023-10-06-multitenancy-in-directus/

**Comment by JPustkuchen** [comment]
Could be called "Spaces" for example, maybe this could be an extension from directus itself?

**Comment by everyx** [comment]
I've implemented a preliminary multi-tenancy support based on `AsyncLocalStorage`, which should have bugs, not fully completed, and I don't plan to continue. However, those interested can use it as a reference: https://github.com/lichitech/directus/tree/feat/multi-tenant
